File: Filtering in MVC 3/[C#]-Filtering in MVC 3/C#/Mvc3Filter/Filters/RequestTimingFilterProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web.Mvc;

public class RequestTimingFilterProvider : IFilterProvider {

    private readonly List<Func<ControllerContext, object>> _conditions = new List<Func<ControllerContext, object>>();
    public void Add(Func<ControllerContext, object> condition) { _conditions.Add(condition); }
    public IEnumerable<Filter> GetFilters(ControllerContext controllerContext, ActionDescriptor actionDescriptor) {
        return from condition in _conditions
               select condition(controllerContext) into filter
               where filter != null
               select new Filter(filter, FilterScope.Global, 1);
    }
}

File: Filtering in MVC 3/[VB]-Filtering in MVC 3/VB/Filters/RequestTimingFilterProvider.vb
Imports System.Web.Mvc

Public Class RequestTimingFilterProvider
	Implements IFilterProvider


	Private ReadOnly _conditions As New List(Of Func(Of ControllerContext, Object))()
	Public Sub Add(ByVal condition As Func(Of ControllerContext, Object))
		_conditions.Add(condition)
	End Sub
    Public Function GetFilters(ByVal controllerContext As ControllerContext, ByVal actionDescriptor As ActionDescriptor) As IEnumerable(Of Filter) Implements IFilterProvider.GetFilters
        Return From condition In _conditions
                 Select filter1 = condition(controllerContext)
                 Where filter1 IsNot Nothing
                 Select New Filter(filter1, FilterScope.Global, 1)

    End Function


End Class